spec/features/saml/saml_spec.rb

```diff
@@ -273,6 +279,240 @@
+
+        visit edit_user_registration_path
+        expect(page).to have_content I18n.t 'devise.failure.unauthenticated'
+      end
+    end
+
+    context 'with multiple SP sessions and SP-initiated logout; non-chronological' do
+      let(:user) { create(:user, :signed_up) }
+      let(:response_xmldoc) { XmlDoc.new('feature', 'response_assertion') }
+      let(:request_xmldoc) { XmlDoc.new('feature', 'request_assertion') }
+
+      before do
+        visit sp1_authnrequest # sp1
+        authenticate_user(user)
+        @sp1_session_index = response_xmldoc.response_session_index_assertion
+        click_button 'Submit'
+
+        visit sp2_authnrequest # sp2
+        @sp2_session_index = response_xmldoc.response_session_index_assertion
+        click_button 'Submit'
+
+        request = OneLogin::RubySaml::Logoutrequest.new
+        settings = sp1_saml_settings
+        settings.name_identifier_value = user.uuid
+        visit request.create(settings) # sp1
+      end
+
+      it 'terminates the user session at IdP' do
+        click_button 'Submit' # LogoutRequest for first SP: sp2
+        click_button 'Submit' # LogoutResponse for originating SP: sp1
+
+        visit edit_user_registration_path
+        expect(page).to have_content I18n.t 'devise.failure.unauthenticated'
+      end
+
+      it 'terminates sessions in order of authentication' do
+        expect(request_xmldoc.asserted_session_index).to eq(@sp2_session_index)
+        click_button 'Submit' # LogoutRequest for first SP: sp2
+
+        expect(response_xmldoc.logout_status_assertion).
+          to eq('urn:oasis:names:tc:SAML:2.0:status:Success')
+        click_button 'Submit' # LogoutResponse for originating SP: sp1
+
+        visit edit_user_registration_path
+        expect(page).to have_content I18n.t 'devise.failure.unauthenticated'
+
+        expect(user.active_identities.size).to eq(0)
+      end
+    end
+
+    context 'with alternate multiple SP sessions and SP-initiated logout; non-chronological' do
+      let(:user) { create(:user, :signed_up) }
+      let(:response_xmldoc) { XmlDoc.new('feature', 'response_assertion') }
+      let(:request_xmldoc) { XmlDoc.new('feature', 'request_assertion') }
+
+      before do
+        visit sp2_authnrequest # sp2
+        authenticate_user(user)
+        @sp2_session_index = response_xmldoc.response_session_index_assertion
+        click_button 'Submit'
+
+        visit sp1_authnrequest # sp1
+        @sp1_session_index = response_xmldoc.response_session_index_assertion
+        click_button 'Submit'
+
+        request = OneLogin::RubySaml::Logoutrequest.new
+        settings = sp2_saml_settings
+        settings.name_identifier_value = user.uuid
+        visit request.create(settings) # sp2
+      end
+
+      it 'terminates the user session at IdP' do
+        find('.btn').click # LogoutRequest for most recent SP: sp1
+        find('.btn').click # LogoutResponse for originating SP: sp2
```

Comment:
Good catch. When I was first working on this PR and trying to get the tests to pass, it couldn't find the Submit button for some reason. That's probably been fixed now. I'll double check and make the change.